File: answers/ruby/problem_004.rb
# Largest palindrome product

## A palindromic number reads the same both ways. The largest palindrome made from the product of two 2-digit numbers is 9009 = 91 × 99.

## Find the largest palindrome made from the product of two 3-digit numbers.

def is_palindromic?(number)
  number_string = number.to_s
  length = number_string.size
  max = (length / 2 - 1) # 0 based index

  (0..max).all? do |index|
    number_string[index] == number_string[-(index+1)]
  end
end

three_digits = (100..999).to_a.reverse
three_digits_all = three_digits.dup

palindromes = []

while three_digits.any?
  check_num = three_digits.shift

  three_digits_all.each do |test_case|
    product = check_num * test_case
    status = is_palindromic?(product)
    palindromes.push(product) if status
  end
end

palindromes.max #=> 906609

File: answers/python3/problem_004.py
# Largest palindrome product

## A palindromic number reads the same both ways. The largest palindrome made from the product of two 2-digit numbers is 9009 = 91 × 99.

## Find the largest palindrome made from the product of two 3-digit numbers.

def is_palindrome(string):
	length = len(string)
	max = int((length / 2 - 1))
	return all(string[i] == string[-(i+1)] for i in range(0, max+1))

three_digits 		= list(range(100, 999+1))
three_digits_all 	= three_digits[:]
max_palindrome 		= 0

while len(three_digits) > 0:
	check_number = three_digits.pop()

	for test_number in three_digits_all:
		product = check_number * test_number
		if is_palindrome(str(product)) and product > max_palindrome:
			max_palindrome = product

print(max_palindrome)
#=> 906609
